src/mito2/src/sst/index/vector_index/creator.rs

```diff
@@ -0,0 +1,873 @@
+        let mut guard = self.stats.record_update();
+        let n = batch.num_rows();
+        guard.inc_row_count(n);
+
+        for (col_id, creator) in &mut self.creators {
+            let Some(values) = batch.field_col_value(*col_id) else {
+                continue;
+            };
+
+            // Process each row in the batch
+            for i in 0..n {
+                let value = values.data.get_ref(i);
+                if value.is_null() {
+                    creator.add_null();
+                } else {
+                    // Extract the vector bytes and convert to f32 slice
+                    if let ValueRef::Binary(bytes) = value {
+                        let floats = bytes_to_f32_slice(bytes);
+                        if floats.len() != creator.config.dim {
+                            return VectorIndexBuildSnafu {
+                                reason: format!(
+                                    "Vector dimension mismatch: expected {}, got {}",
+                                    creator.config.dim,
+                                    floats.len()
+                                ),
+                            }
+                            .fail();
+                        }
+                        creator.add_vector(&floats)?;
+                    } else {
+                        creator.add_null();
+                    }
+                }
+            }
+
+            // Check memory limit - abort index creation if exceeded
+            if let Some(threshold) = self.memory_usage_threshold {
+                let current_usage = creator.memory_usage();
+                if current_usage > threshold {
+                    warn!(
+                        "Vector index memory usage {} exceeds threshold {}, aborting index creation",
+                        current_usage, threshold
+                    );
+                    return VectorIndexBuildSnafu {
+                        reason: format!(
+                            "Memory usage {} exceeds threshold {}",
+                            current_usage, threshold
+                        ),
+                    }
+                    .fail();
+                }
+            }
+        }
+
+        Ok(())
+    }
+
+    /// Internal flat update implementation.
+    async fn do_update_flat(&mut self, batch: &RecordBatch) -> Result<()> {
+        let mut guard = self.stats.record_update();
+        let n = batch.num_rows();
+        guard.inc_row_count(n);
+
+        for (col_id, creator) in &mut self.creators {
+            let Some(column_meta) = self.metadata.column_by_id(*col_id) else {
+                for _ in 0..n {
+                    creator.add_null();
+                }
+                continue;
+            };
+
+            let column_name = &column_meta.column_schema.name;
+            let Some(column_array) = batch.column_by_name(column_name) else {
+                for _ in 0..n {
+                    creator.add_null();
```

Comment:
nit: maaybe add a `add_nulls(n: usize)` to creator for simplicity and efficiency

Reply:
Addressed in https://github.com/GreptimeTeam/greptimedb/pull/7468/changes/2eda47826a60582e7686e00ade36fecdda10eb01